### PR TITLE
[DO_NOT_MERGE] Add JShell support

### PR DESCRIPTION
**Pull request by hchoura**
@nanoscott

### CODE REVIEW

File: src/foam/nanos/script/Script.js [hunk 1]
```diff
@@ -201,28 +239,66 @@ foam.CLASS({
...
+              l1.add(line);
+            }
+
+            List<String> instractionList = new InstractionPresentation(jShell).parseToInstraction(l1);
+
+            EvalInstraction console = new EvalInstraction(jShell, instractionList, x);
+            print = console.runEvalInstraction();
+            ps.print(print);
+          } catch (Throwable e) {
+            e.printStackTrace();
```
Comment (jlhughes) [inline src/foam/nanos/script/Script.js:269]: Is it possible to log this?

File: src/foam/nanos/script/Script.js [hunk 2]
```diff
@@ -201,28 +239,66 @@ foam.CLASS({
...
+
+          // TODO: ',common packages like foam.core.*, foam.dao.*, etc.
+          try {
+            setOutput("");
+            shell.setOut(ps);
+            shell.eval(getCode());
+          } catch (Throwable e) {
+            ps.println();
+            e.printStackTrace(ps);
+            e.printStackTrace();
```
Comment (jlhughes) [inline src/foam/nanos/script/Script.js:292]: Is it possible to log this?

File: src/foam/nanos/script/jShell/EvalInstraction.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ */
+package foam.nanos.script.jShell;
+
+import java.util.List;
+
+import foam.core.X;
+import jdk.jshell.JShell;
+import jdk.jshell.SnippetEvent;
+
+public class EvalInstraction {
```
Comment (jlhughes) [inline src/foam/nanos/script/jShell/EvalInstraction.java:14]: Is Instraction a typo? Should it be Instruction?

File: src/foam/nanos/script/jShell/InstractionPresentation.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ */
+package foam.nanos.script.jShell;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import jdk.jshell.JShell;
+import jdk.jshell.SourceCodeAnalysis;
+
+public class InstractionPresentation {
```
Comment (jlhughes) [inline src/foam/nanos/script/jShell/InstractionPresentation.java:14]: Is this a typo 'Instraction'?

File: src/foam/nanos/script/jShell/ReadLineByLine.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  public List<String> ReadByLine(String initialScript) {
+    try {
+      BufferedReader br = new BufferedReader(new FileReader(initialScript));
+      for ( String scriptLine; ( scriptLine = br.readLine() ) != null; ) {
+        if ( SafetyUtil.isEmpty(scriptLine.trim()) ) continue;
+        this.lineScripts.add(scriptLine);
+      }
+      br.close();
+    } catch (Exception e) {
+      e.printStackTrace();
```
Comment (jlhughes) [inline src/foam/nanos/script/jShell/ReadLineByLine.java:36]: Log please, if possible.

File: src/foam/nanos/boot/NSpec.js [hunk 6]
```diff
@@ -170,8 +213,14 @@ foam.CLASS({
...
       readPermissionRequired: true,
       writePermissionRequired: true
-    }
+    },
     // TODO: permissions, keywords, lazy, parent
+    {
+      class: 'Object',
+      name: 'objectJshell',
+      static: true
+      //javaValue: 'new Object[1]'
```
Comment (nanoscott) [inline src/foam/nanos/boot/NSpec.js:222]: remove dead code.

File: src/foam/nanos/boot/NSpec.js [hunk 7]
```diff
@@ -207,20 +257,61 @@ foam.CLASS({
...
+              String init = "import foam.core.X; import foam.nanos.boot.NSpec; X x = foam.nanos.script.Script.X_HOLDER[0]; ";
+              
+              BufferedReader rdr = new BufferedReader(new StringReader(init+getServiceScript()));
+              List<String> l1 = new ArrayList<String>();
+  
+              for ( String line = rdr.readLine(); line != null; line = rdr.readLine() ) {
+                l1.add(line);
+              }
+  
+              List<String> instractionList = new InstractionPresentation(jShell).parseToInstraction(l1);
```
Comment (nanoscott) [inline src/foam/nanos/boot/NSpec.js:290]:
```suggestion
              List<String> instructionList = new InstructionPresentation(jShell).parseToInstruction(l1);
```

File: src/foam/nanos/boot/NSpec.js [hunk 8]
```diff
@@ -207,20 +257,61 @@ foam.CLASS({
...
+              EvalInstraction console = new EvalInstraction(jShell, instractionList, x);
```
Comment (nanoscott) [inline src/foam/nanos/boot/NSpec.js:291]:
```suggestion
              EvalInstruction console = new EvalInstruction(jShell, instructionList, x);
```

File: src/foam/nanos/boot/NSpec.js [hunk 9]
```diff
@@ -207,20 +257,61 @@ foam.CLASS({
...
+              String print = null;
+              print = console.runEvalInstraction();
```
Comment (nanoscott) [inline src/foam/nanos/boot/NSpec.js:293]:
```suggestion
              print = console.runEvalInstruction();
```

File: src/foam/nanos/boot/NSpec.js [hunk 10]
```diff
@@ -207,20 +257,61 @@ foam.CLASS({
...
+
+//          saveService(x, service)//TODO delete this method.
```
Comment (nanoscott) [inline src/foam/nanos/boot/NSpec.js:295]: remove dead code.

File: src/foam/nanos/boot/NSpec.js [hunk 11]
```diff
@@ -229,7 +320,11 @@ foam.CLASS({
       javaThrows: [
         'java.lang.ClassNotFoundException',
         'java.lang.InstantiationException',
-        'java.lang.IllegalAccessException'
+        'java.lang.IllegalAccessException',
+        'SecurityException',
+        'NoSuchFieldException',//IllegalArgumentException,
```
Comment (nanoscott) [inline src/foam/nanos/boot/NSpec.js:325]: remove the dead code here

File: src/foam/nanos/boot/NSpecFactory.java [hunk 12]
```diff
@@ -45,7 +48,14 @@ public synchronized Object create(X x) {
 
     try {
       if ( logger != null ) logger.info("Creating Service", spec_.getName());
-      ns_ = spec_.createService(x_.getX().put(NSpec.class, spec_));
+      if ( spec_.getLanguage() == foam.nanos.script.Language.BEANSHELL ) {
+        ns_ = spec_.createService(x_.getX().put(NSpec.class, spec_),null);
+      }
+      else if (spec_.getLanguage() == foam.nanos.script.Language.JSHELL) {
+        //PrintStream ps =null;
```
Comment (nanoscott) [inline src/foam/nanos/boot/NSpecFactory.java:55]: remove dead code

File: src/foam/nanos/boot/NSpecFactory.java [hunk 13]
```diff
@@ -45,7 +48,14 @@ public synchronized Object create(X x) {
...
+        ns_ = spec_.createService(x_.getX().put(NSpec.class, spec_),null);
```
Comment (nanoscott) [inline src/foam/nanos/boot/NSpecFactory.java:56]: there doesn't seem to be a difference between this line and the line above for beanshell .. so why differentiate between the two?

File: src/foam/nanos/script/Script.js [hunk 14]
```diff
@@ -164,21 +172,51 @@ foam.CLASS({
...
+          Interpreter shell = new Interpreter();
+
+          try {
+            shell.set("currentScript", this);
+            shell.set("x", x);
+            shell.eval(
+                "runScript(String name) { script = x.get(\\"scriptDAO\\").find(name); if ( script != null ) eval(script.code); }");
+            shell.eval("foam.core.X sudo(String user) { foam.util.Auth.sudo(x, (String) user); }");
+            shell.eval("foam.core.X sudo(Object id) { foam.util.Auth.sudo(x, id); }");
+          } catch (EvalError e) {
```
Comment (nanoscott) [inline src/foam/nanos/script/Script.js:214]: log the error.

**Pull request closed without merging**